core/src/main/java/com/spotify/metrics/core/SemanticMetricRegistryListenerV2.java

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright (c) 2016 Spotify AB.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.metrics.core;
+
+import com.spotify.metrics.core.codahale.metrics.ext.Distribution;
+
+public interface SemanticMetricRegistryListenerV2 extends SemanticMetricRegistryListener {
```

Comment:
I'm curious why do we need this? I've checked GHE and Github and don't see any usage of SemanticMetricRegistryListener. Maybe we could try to avoid introducing V2 and just extend it with additional methods?

Reply:
I'll also note that Version numbers in class names is quite atypical. Most often a new xyzv2 module is created and the new incompatible version is put in that. Then you can switch interfaces just by changing import statements.

Reply:
SematicMetricRegistryListener is used in  SemanticRegistry. It is also part of  SemanticRegistry public interface so we need this update. Please check this method https://github.com/spotify/semantic-metrics/blob/master/core/src/main/java/com/spotify/metrics/core/SemanticMetricRegistry.java#L256.


As of the name, I didn't rely on the package name in this case because it can easily lead to a bug.  You have to look at the import statement or explicitly add the package name when you refer to it.   I can change the name to SemanticMetricRegistryListenerWithDistribution

Reply:
Could you elaborate why we couldn't extend SematicMetricRegistryListener with 2 new methods? Why do we need to create new SematicMetricRegistryListenerV2 extending SematicMetricRegistryListener and adding new methods?

Reply:
If we add two methods instead of extending it, anyone using SemacticMetricRegistryListener will have to make code change whether you want to use distribution or not. 
That goes against the idea of opt-in for distribution.

Reply:
I think the problem is if a third party does:

```java
class X extends SemanticMetricRegistryListener {
}
```

and then upgrade the version of semantic-metrics, class X will crash in runtime since it does not implement the two new versions.

That's why you would need to either create a new interface (like is done in this PR) or let the new methods have a default implementation.

Personally I would prefer adding these as default methods with an empty method body instead of creating a new subinterface.

Reply:
I would agree here and that's exactly what I had in mind - having default methods and keeping the name as is.

Reply:
@malish8632 you cannot override default method. What is a point of having an interface that implements a listener if you cannot override the methods.

Reply:
> I think the simplest thing to do is to to add new methods to the `SemanticMetricRegistryListener` interface and bump the major version number of semantic-metrics.
> 
> You're already breaking the API of people using `SemanticMetricRegistry.addListener(SemanticMetricRegistryListener)`.
> 
> if I have a class that implements `SemanticMetricRegistryListener`, I probably also want to update it for the new methods (I've done this before when implementing some sort of bridge between semantic-metrics and other libraries, for instance)
@mattnworb that will lead to the problem one reviewer mentioned above. Also we decided that Distribution f will be an opt-in feature.

Reply:
I missed @spkrka's comment, that is a good point. But you can definitely override a default method, thats what makes them _default_ 😀 

what does it mean for it to be an opt-in feature, aren't all these metric types opt-in? I don't have to use them in my application code if I don't have a need for them.

Reply:
hey @ao2017 I think you can and that's the point of that method to be used only if you don't override it. [See here.](https://docs.oracle.com/javase/tutorial/java/IandI/defaultmethods.html)

Reply:
Ok, you can override any method as long as it is not final or private :). With the default method the compiler doesn't enforce the need for override.  Anyway, since there is a strong preference for this I removed the new interface.